src/test/java/teammates/storage/api/NotificationsDbTest.java

```diff
@@ -0,0 +1,251 @@
+package teammates.storage.api;
+
+import java.time.Instant;
+import java.util.UUID;
+
+import org.testng.annotations.Test;
+
+import teammates.common.datatransfer.NotificationTargetUser;
+import teammates.common.datatransfer.NotificationType;
+import teammates.common.datatransfer.attributes.NotificationAttributes;
+import teammates.common.exception.EntityAlreadyExistsException;
+import teammates.common.exception.InvalidParametersException;
+import teammates.storage.entity.Notification;
+import teammates.test.BaseTestCaseWithLocalDatabaseAccess;
+
+/**
+ * SUT: {@link NotificationsDb}.
+ */
+public class NotificationsDbTest extends BaseTestCaseWithLocalDatabaseAccess {
+
+    private final NotificationsDb notificationsDb = NotificationsDb.inst();
+
+    @Test
+    public void testGetNotification() throws Exception {
+        NotificationAttributes n = createNewNotification();
+
+        ______TS("typical success case");
+        NotificationAttributes retrieved = notificationsDb.getNotification(n.getNotificationId());
+        assertNotNull(retrieved);
+
+        ______TS("expect null for non-existent account");
+        retrieved = notificationsDb.getNotification("non.existent");
+        assertNull(retrieved);
+
+        ______TS("failure: null parameter");
+        assertThrows(AssertionError.class, () -> notificationsDb.getNotification(null));
+
+        // delete created account
+        notificationsDb.deleteNotification(n.getNotificationId());
+    }
+
+    @Test
+    public void testCreateNotification() throws Exception {
+        ______TS("typical success case");
+        String uuid = UUID.randomUUID().toString();
+        NotificationAttributes n = NotificationAttributes.builder(uuid)
+                .withTitle("title")
+                .withMessage("demo message")
+                .withType(NotificationType.DEPRECATION)
+                .withTargetUser(NotificationTargetUser.GENERAL)
+                .withStartTime(Instant.now())
+                .withEndTime(Instant.now().plusSeconds(3600))
+                .build();
+
+        notificationsDb.createEntity(n);
+
+        ______TS("duplicate notification, creation fail");
+        NotificationAttributes dup = NotificationAttributes.builder(uuid)
+                .withTitle("title2")
+                .withMessage("demo message2")
+                .withType(NotificationType.MAINTENANCE)
+                .withTargetUser(NotificationTargetUser.STUDENT)
+                .withStartTime(Instant.now())
+                .withEndTime(Instant.now().plusSeconds(3600))
+                .build();
+
+        assertThrows(EntityAlreadyExistsException.class, () -> notificationsDb.createEntity(dup));
+
+        ______TS("failure: null parameters");
+        assertThrows(AssertionError.class, () -> notificationsDb.createEntity(null));
+    }
```

Comment:
Update/Create method will not fail at the Db level. For the invalid parameters, checks are done at the Attributes class so shouldn't that be under AttributeTest (PR has not been created yet), not DbTest?

Reply:
You are right that the attribute class performs the checks but it is the Db class that actually throws the error if the data is invalid, see the `EntitiesDb` class. The reason why we need to test that `InvalidParametersException` is thrown is because of that.

Reply:
Btw! `NotificationAttributesTest` is already being added by @ziqing26  https://github.com/TEAMMATES/teammates/pull/11648